httpcore5/src/main/java/org/apache/hc/core5/http/protocol/ViaRequest.java

```diff
@@ -0,0 +1,127 @@
+ * only
+ * if it has not been added previously, as per the HTTP specification. Additionally, it updates the
+ * values in the {@link HttpHeaders#VIA} header correctly in case of multiple intermediate protocols
+ * or recipients, by appending its own information about how the message was received to the end of
+ * the header field value.
+ *
+ * @since 5.3
+ */
+@Contract(threading = ThreadingBehavior.IMMUTABLE)
+public class ViaRequest implements HttpRequestInterceptor {
+
+
+    /**
+     * Singleton instance.
+     */
+    public static final HttpRequestInterceptor INSTANCE = new ViaRequest();
+
+    /**
+     * Constructs a new {@code ViaRequest}.
+     */
+    public ViaRequest() {
+    }
+
+    /**
+     * Adds the HTTP {@link  HttpHeaders#VIA} header to the request if it does not already exist.
+     *
+     * <p>This method ensures that the version of the request is HTTP/1.1 or higher, and adds the
+     * Via header in the format {@code <protocol> <version> <host>}, where {@code <protocol>} is the protocol name,
+     * {@code <version>} is the major and minor version of the request, and {@code <host>} is the value of the Host header.
+     *
+     * <p>In case the {@link  HttpHeaders#VIA} header already exists, this method updates its value by appending
+     * the new protocol information in the same format.
+     *
+     * <p>If the version of the request is lower than {@code HTTP/1.1} or the request authority not being specified,
+     * this method throws a {@link ProtocolException}.
+     *
+     * @param request the request object to modify
+     * @param entity the entity for the request, may be {@code null}
+     * @param context the context for the request
+     * @throws ProtocolException if there was a protocol error, such as the request version being lower than {@code HTTP/1.1},
+     *         or the request authority not being specified
+     * @throws IOException if there was an I/O error
+     */
+    @Override
+    public void process(final HttpRequest request, final EntityDetails entity, final HttpContext context) throws ProtocolException, IOException {
+        Args.notNull(request, "HTTP request");
+        Args.notNull(context, "HTTP context");
+        final ProtocolVersion ver = context.getProtocolVersion() != null ? context.getProtocolVersion() : HttpVersion.HTTP_1_1;
+
+        final URIAuthority authority = request.getAuthority();
+        if (authority == null) {
+            throw new ProtocolException("Request authority not specified");
+        }
+
+
+       if (!ver.greaterEquals(HttpVersion.HTTP_1_1)) {
+           throw new ProtocolException("Invalid protocol version: %s", ver);
+       }
+       String viaHeaderValue = ver.getProtocol() + " " + ver.getMajor() + "." + ver.getMinor() + " " + authority.getHostName();
+       final int port = authority.getPort();
+       if (port != -1) {
+            viaHeaderValue += ":" + port;
+        }
+        if (!request.containsHeader(HttpHeaders.VIA)) {
```

Comment:
@arturobernalg One last thing. Please move the `Via` header value generation into this if statement. There is not point going through the whole process just to find out the value is not needed after all.

Reply:
@ok2c 
Something like this
` if (!request.containsHeader(HttpHeaders.VIA)) {
            String viaHeaderValue = ver.getProtocol() + " " + ver.getMajor() + "." + ver.getMinor() + " " + authority.getHostName();
            final int port = authority.getPort();
            if (port != -1) {
                viaHeaderValue += ":" + port;
            }
            request.addHeader(HttpHeaders.VIA, viaHeaderValue);
        }`

Reply:
@arturobernalg Something like that.